### PR TITLE
Preserve eltype where possible for moments

### LINKED ISSUE #680: std(::AbstractVector{Float32}, ::AnalyticWeights{Float32}) gives a Float64
Setup:
```julia
julia> using StatsBase, Statistics

julia> xs = Float32[1, 2, 3, 4, 5];

julia> ws = AnalyticWeights(Float32[1, 1, 1, 1, 1]);

julia> eltype(ws)
Float32
```

Now trying some operations

```julia
julia> typeof(std(xs))  # As expected
Float32


julia> typeof(mean(xs, ws))  # as expected
Float32

julia> typeof(std(xs, ws))  # Not expected Should be a Float32
Float64
```

I was pretty sure this used to work to give a `Float32`
if so, I can't work-out when it changed.

### PR DESCRIPTION
**Pull request by palday**
Fixes #680.

### CODE REVIEW

**Review by oxinabox — APPROVED** [review]

File: src/moments.jl [hunk 1]
```diff
@@ -221,19 +221,19 @@ end
 
 
 ##### General central moment
-function _moment2(v::RealArray, m::Real; corrected=false)
+function _moment2(v::RealArray{T}, m::Real; corrected=false) where T
```
Comment (ararslan) [inline src/moments.jl:224]: Rather than forcing specialization, I would just replace the use of `T` below with `eltype(v)`. Same for all of the other methods here.

File: src/moments.jl [hunk 2]
```diff
@@ -242,59 +242,59 @@ function _moment2(v::RealArray, wv::AbstractWeights, m::Real; corrected=false)
...
+    s = zero(T)
     for i = 1:n
         @inbounds z = v[i] - m
         s += z * z * z
     end
     s / n
 end
 
-function _moment3(v::RealArray, wv::AbstractWeights, m::Real)
+function _moment3(v::RealArray{T}, wv::AbstractWeights, m::Real; corrected=false) where T
```
Comment (ararslan) [inline src/moments.jl:255]: The added keyword argument here seems not to be used?

File: src/moments.jl [hunk 3]
```diff
@@ -419,7 +419,7 @@ function kurtosis(v::RealArray, wv::AbstractWeights, m::Real)
     sw = sum(wv)
     cm4 /= sw
     cm2 /= sw
-    return (cm4 / (cm2 * cm2)) - 3.0
+    return (cm4 / (cm2 * cm2)) - 3 * one(T)
```
Comment (ararslan) [inline src/moments.jl:422]:
```suggestion
    return (cm4 / (cm2 * cm2)) - 3
```

File: src/moments.jl (hunk 1 above)
Comment (mschauer) [reply src/moments.jl:224]: “Forcing specialization”? Isn’t Julia always specializing?

File: src/moments.jl (hunk 2 above)
Comment (palday) [reply src/moments.jl:255]: copy-paste error

File: src/moments.jl (hunk 1 above)
Comment (palday) [reply src/moments.jl:224]: @ararslan that's how I originally had it, but I wanted to get some additional type stability because I figured that moments might be something in a tight loop. It turned out harder to have that specialization on the weight vector (which is why it's not specialized) because the type parameterization is complicated and nested.

@mschauer I think the trick here is that it's parametric types and how much specialization is done on the type parameter.  

I'm happy to have the simpler variant with `eltype`.

**Comment by palday** [comment]
The failure on nightly is an upstream bug (https://github.com/JuliaLang/julia/issues/40609).

File: src/moments.jl [hunk 4]
```diff
@@ -223,7 +223,7 @@ end
 ##### General central moment
 function _moment2(v::RealArray, m::Real; corrected=false)
     n = length(v)
-    s = 0.0
+    s = zero(eltype(v))
```
Comment (ararslan) [inline src/moments.jl:226]:
```suggestion
    s = zero(promote_type(eltype(v), typeof(m))
```

Comment (ararslan) [reply src/moments.jl:226]: I think this applies for all of changes here.

File: src/moments.jl [hunk 5]
```diff
@@ -264,17 +264,17 @@ end
...
     for i = 1:n
         @inbounds z = v[i] - m
         s += abs2(z * z)
     end
     s / n
 end
 
-function _moment4(v::RealArray, wv::AbstractWeights, m::Real)
+function _moment4(v::RealArray, wv::AbstractWeights, m::Real; corrected=false)
```
Comment (ararslan) [inline src/moments.jl:275]:
```suggestion
function _moment4(v::RealArray, wv::AbstractWeights, m::Real)
```

File: src/moments.jl (hunk 4 above)
Comment (palday) [reply src/moments.jl]: Everybody wants a promotion...

File: src/moments.jl [hunk 6]
```diff
@@ -85,14 +85,16 @@ end
 function varm(A::RealArray, w::AbstractWeights, M::RealArray, dim::Int;
                    corrected::DepBool=nothing)
     corrected = depcheck(:varm, corrected)
-    varm!(similar(A, Float64, Base.reduced_indices(axes(A), dim)), A, w, M,
+    T = promote_type(promote_type(eltype(A), eltype(w)), eltype(M))
```
Comment (ararslan) [inline src/moments.jl:88]: `promote_type` has a vararg method, no need to nest calls

Comment (palday) [reply src/moments.jl:88]: If only that were in its docstring :angry:

**Comment by ararslan** [comment]
This looks good but the prior uses of `Float64` make me worried that the code may have issues around under/overflow when using other types.

**Comment by palday** [comment]
Ugh, I'm conflicted on that front. Because that would be breaking, but technically a user hitting overflow on their custom narrow types is expected behavior. @nalimilan do we have enough for another breaking pre-1.0 release?

**Comment by nalimilan** [comment]
We can break things if we want to tag 1.0. Though ideally I'd rather move these to Statistics, but that work has stalled (https://github.com/JuliaLang/julia/pull/31395).

File: src/moments.jl [hunk 7]
```diff
@@ -233,7 +236,8 @@ end
 
 function _moment2(v::RealArray, wv::AbstractWeights, m::Real; corrected=false)
     n = length(v)
-    s = 0.0
+    T = promote_type(eltype(v), eltype(wv), typeof(m))
+    s = zero(T)
```
Comment (nalimilan) [inline src/moments.jl:240]: To ensure type stability, how about following exactly what the loop does and using `T = typeof(zero(eltype(v) - zero(m))^2 * zero(eltype(w)))`? That's what I did in my first attempt to moving these functions to Statistics: https://github.com/JuliaLang/julia/pull/31395/files#diff-6b82a0c4e60cbe68048e35375987a86688afb4aea60dda8f2d2d4938dc845776R375

**Comment by palday** [comment]
Okay, so on the breaking / non breaking front:

1. Did the API ever specify the return type for `mean` and the various moments? If not, this is not breaking in terms of type behavior.
2. In terms of numeric behavior, Julia generally keeps things at narrow types even at the risk of over-/underflow. As such, keeping the narrowest type that would be possible for the arithmetic operations in these computations seems fine. In other words, the new return type and potential numeric implications are expected in the Julia ecosystem. 
3. The potential numeric issues with all of the moments here are possible regardless of type width -- we do accumulate then divide without sorting first, so
    - we have issues with computations where things differ in scale and the smaller values accumulate enough to be relevant on the larger scale, but consistently round down when the initial values in the array are large (sorting solves this but is computationally expensive)
    - we can have overflow in the accumulation step (batching the accumulate+divide on subsets solves this but potentially causes performance and numeric issues from the extra divisions)
4. The way for best numerical accuracy would be sort, do the math in a higher precision and cast down. But this also seems unacceptable -- if a user choose to keep everything in single or half precision, then we should assume there is a reason for that. If they really need that level of numerical stability, then they should probably just implement their own moments.

In other words: given that we're not going to change the algorithm used (which I think is reasonable), the potential numeric issues from the eltype preservation should not be considered breaking. Perhaps the best solution is to add to each of the docstrings:

```markdown
!!! note
    The computation is done in the narrowest type for which the underlying 
    arithmetic operations are defined based on the element types of the 
    vectors and the mean. For higher numerical accuracy and to reduce the risk
    of overflow when dealing with values near the limits of the type (or many 
    values whose sum is near the limits of the type), we recommend casting to 
    a broader type / higher precision first.
```

File: src/moments.jl [hunk 8]
```diff
@@ -281,8 +276,7 @@ end
 
 function _moment4(v::RealArray, wv::AbstractWeights, m::Real)
     n = length(v)
-    T = promote_type(eltype(v), eltype(wv), typeof(m))
-    s = zero(T)
+    s = zero(eltype(wv))  * (zero(eltype(v)) - zero(m))^2
```
Comment (nalimilan) [inline src/moments.jl:279]:
```suggestion
    s = zero(eltype(wv)) * (zero(eltype(v)) - zero(m))^4
```

File: src/moments.jl [hunk 9]
```diff
@@ -236,8 +235,7 @@ end
 
 function _moment2(v::RealArray, wv::AbstractWeights, m::Real; corrected=false)
     n = length(v)
-    T = promote_type(eltype(v), eltype(wv), typeof(m))
-    s = zero(T)
+    s = zero(eltype(wv))  * (zero(eltype(v)) - zero(m))^2
```
Comment (nalimilan) [inline src/moments.jl:238]:
```suggestion
    s = zero(eltype(wv)) * (zero(eltype(v)) - zero(m))^2
```

File: src/moments.jl [hunk 10]
```diff
@@ -248,8 +246,7 @@ end
 
 function _moment3(v::RealArray, m::Real)
     n = length(v)
-    T = promote_type(eltype(v), typeof(m))
-    s = zero(T)
+    s = zero(eltype(v)) - zero(m)
```
Comment (nalimilan) [inline src/moments.jl:249]:
```suggestion
    s = (zero(eltype(v)) - zero(m))^3
```

File: src/moments.jl [hunk 11]
```diff
@@ -225,8 +225,7 @@ end
 ##### General central moment
 function _moment2(v::RealArray, m::Real; corrected=false)
     n = length(v)
-    T = promote_type(eltype(v), typeof(m))
-    s = zero(T)
+    s = zero(eltype(v)) - zero(m)
```
Comment (nalimilan) [inline src/moments.jl:228]:
```suggestion
    s = (zero(eltype(v)) - zero(m))^2
```

File: src/moments.jl [hunk 12]
```diff
@@ -259,8 +256,7 @@ end
 
 function _moment3(v::RealArray, wv::AbstractWeights, m::Real)
     n = length(v)
-    T = promote_type(eltype(v), eltype(wv), typeof(m))
-    s = zero(T)
+    s = zero(eltype(wv))  * (zero(eltype(v)) - zero(m))^2
```
Comment (nalimilan) [inline src/moments.jl:259]:
```suggestion
    s = zero(eltype(wv)) * (zero(eltype(v)) - zero(m))^3
```

File: src/moments.jl [hunk 13]
```diff
@@ -270,8 +266,7 @@ end
 
 function _moment4(v::RealArray, m::Real)
     n = length(v)
-    T = promote_type(eltype(v), typeof(m))
-    s = zero(T)
+    s = zero(eltype(v)) - zero(m)
```
Comment (nalimilan) [inline src/moments.jl:269]:
```suggestion
    s = (zero(eltype(v)) - zero(m))^4
```

File: src/moments.jl [hunk 14]
```diff
@@ -303,8 +296,7 @@ end
 
 function _momentk(v::RealArray, k::Int, wv::AbstractWeights, m::Real)
     n = length(v)
-    T = promote_type(eltype(v), eltype(wv), typeof(m))
-    s = zero(T)
+    s = zero(eltype(wv))  * (zero(eltype(v)) - zero(m))^2
```
Comment (nalimilan) [inline src/moments.jl:299]:
```suggestion
    s = zero(eltype(wv)) * (zero(eltype(v)) - zero(m))^k
```

File: src/moments.jl [hunk 15]
```diff
@@ -85,14 +85,16 @@ end
 function varm(A::RealArray, w::AbstractWeights, M::RealArray, dim::Int;
                    corrected::DepBool=nothing)
     corrected = depcheck(:varm, corrected)
-    varm!(similar(A, Float64, Base.reduced_indices(axes(A), dim)), A, w, M,
+    T = promote_type(eltype(A), eltype(w), eltype(M))
```
Comment (nalimilan) [inline src/moments.jl:88]: Likewise, here and below it should be possible to use `zero` to choose the appropriate type? That's not supported currently given the signature, but that would allow e.g. computing the standard deviation of Unitful types.

**Comment by nalimilan** [comment]
Yeah it's not super likely to break user code. Though we could collect the few changes which are breaking (some of them are marked with the [breaking](https://github.com/JuliaStats/StatsBase.jl/labels/breaking) label) and take that occasion to tag 1.0.

File: src/moments.jl [hunk 16]
```diff
@@ -341,8 +348,9 @@ specifying a weighting vector `wv` and a center `m`.
 """
 function skewness(v::RealArray, m::Real)
     n = length(v)
-    cm2 = 0.0   # empirical 2nd centered moment (variance)
-    cm3 = 0.0   # empirical 3rd centered moment
+    T = promote_type(eltype(v), typeof(m))
```
Comment (nalimilan) [inline src/moments.jl:351]: Use the same strategy here and below too?
